src/core/vector/qgsvectorlayereditbuffer.cpp

```diff
@@ -338,53 +348,16 @@ bool QgsVectorLayerEditBuffer::commitChanges( QStringList &commitErrors )
   // no                 yes                   => changeAttributeValues
   // yes                yes                   => changeFeatures
 
+  // Check geometry types
   // to fix https://github.com/qgis/QGIS/issues/23663
   if ( !mAddedFeatures.isEmpty() )
-  {
-    if ( cap & QgsVectorDataProvider::AddFeatures )
-    {
-      if ( provider->doesStrictFeatureTypeCheck() )
-      {
-        for ( const QgsFeature &f : std::as_const( mAddedFeatures ) )
-        {
-          if ( ( ! f.hasGeometry() ) ||
-               ( f.geometry().wkbType() == provider->wkbType() ) )
-            continue;
-
-          if ( provider->convertToProviderType( f.geometry() ).isNull() )
-          {
-            commitErrors << tr( "ERROR: %n feature(s) not added - geometry type is not compatible with the current layer.", "not added features count", mAddedFeatures.size() );
-            success = false;
-            break;
-          }
-        }
-      }
-    }
-    else
-    {
-      commitErrors << tr( "ERROR: %n feature(s) not added - provider doesn't support adding features.", "not added features count", mAddedFeatures.size() );
-      success = false;
-    }
-  }
+    success &= commitChangesCheckGeometryTypeCompatibility( commitErrors );
 
   //
   // update geometries
   //
-  if ( success && !mChangedGeometries.isEmpty() && ( ( cap & QgsVectorDataProvider::ChangeFeatures ) == 0 || mChangedAttributeValues.isEmpty() ) )
-  {
-    if ( provider->changeGeometryValues( mChangedGeometries ) )
-    {
-      commitErrors << tr( "SUCCESS: %n geometries were changed.", "changed geometries count", mChangedGeometries.size() );
-
-      emit committedGeometriesChanges( L->id(), mChangedGeometries );
-      mChangedGeometries.clear();
-    }
-    else
-    {
-      commitErrors << tr( "ERROR: %n geometries not changed.", "not changed geometries count", mChangedGeometries.size() );
-      success = false;
-    }
-  }
+  if ( success && !mChangedGeometries.isEmpty() )
+    success &= commitChangesUpdateGeometry( commitErrors );
```

Comment:
There is a difference to the proposal here, that geometries are changed as first step, whereas in the proposal they are changed next to attributes. I guess this is the first step for historical reasons and would propose to clean that up and move it to later.

Reply:
I moved the update of the geometries after the update of attributes as described in the proposal